examples/nlp/multimodal_entailment.py

```diff
@@ -0,0 +1,577 @@
+row = train_df.iloc[idx]
+sample_text_1, sample_text_2 = row["text_1"], row["text_2"]
+print(f"Text 1: {sample_text_1}")
+print(f"Text 2: {sample_text_2}")
+
+test_text = [np.array([sample_text_1]), np.array([sample_text_2])]
+text_preprocessed = bert_preprocess_model(test_text)
+
+print("Keys           : ", list(text_preprocessed.keys()))
+print("Shape Word Ids : ", text_preprocessed["input_word_ids"].shape)
+print("Word Ids       : ", text_preprocessed["input_word_ids"][0, :16])
+print("Shape Mask     : ", text_preprocessed["input_mask"].shape)
+print("Input Mask     : ", text_preprocessed["input_mask"][0, :16])
+print("Shape Type Ids : ", text_preprocessed["input_type_ids"].shape)
+print("Type Ids       : ", text_preprocessed["input_type_ids"][0, :16])
+
+
+"""
+We will now create `tf.data.Dataset` objects from the dataframes for performance.
+
+Note that the text inputs will be preprocessed as a part of the data input pipeline. But
+the preprocessing modules can also be a part of their corresponding BERT models. This
+helps reduce the training/serving skew and lets our models operate with raw text inputs.
+Follow [this tutorial](https://www.tensorflow.org/text/tutorials/classify_text_with_bert)
+to know more about how to incorporate the preprocessing modules directly inside the
+models. 
+"""
+
+
+def dataframe_to_dataset(dataframe):
+    columns = ["image_1_path", "image_2_path", "text_1", "text_2", "label_idx"]
+    dataframe = dataframe[columns].copy()
+    labels = dataframe.pop("label_idx")
+    ds = tf.data.Dataset.from_tensor_slices((dict(dataframe), labels))
+    ds = ds.shuffle(buffer_size=len(dataframe))
+    return ds
+
+
+"""
+### Preprocessing utilities
+"""
+
+resize = (128, 128)
+bert_input_features = ["input_word_ids", "input_type_ids", "input_mask"]
+
+
+def read_resize(image_path):
+    extension = tf.strings.split(image_path)[-1]
+
+    image = tf.io.read_file(image_path)
+    if extension == b"jpg":
+        image = tf.image.decode_jpeg(image, 3)
+    else:
+        image = tf.image.decode_png(image, 3)
+    image = tf.image.resize(image, resize)
+    return image
+
+
+def preprocess_text(text_1, text_2):
+    text_1 = tf.convert_to_tensor([text_1])
+    text_2 = tf.convert_to_tensor([text_2])
+    output = bert_preprocess_model([text_1, text_2])
+    output = {feature: tf.squeeze(output[feature]) for feature in bert_input_features}
+    return output
+
+
+def preprocess(sample):
+    image_1 = read_resize(sample["image_1_path"])
+    image_2 = read_resize(sample["image_2_path"])
+    text = preprocess_text(sample["text_1"], sample["text_2"])
+    return {"image_1": image_1, "image_2": image_2, "text": text}
+
+
+"""
+### Create the final datasets
+"""
+
+batch_size = 32
+auto = tf.data.AUTOTUNE
+
+
+def prepare_dataset(dataframe, training=True):
+    ds = dataframe_to_dataset(dataframe)
+    if training:
+        ds = ds.shuffle(len(train_df))
+    ds = ds.map(lambda x, y: (preprocess(x), y))
```

Comment:
You should add caching to avoid doing the same preprocessing multiple times. You'll already buffering the data in memory (since you do a global shuffle) so this is free. `.cache()`